In [ ]:
pip install pandas

In [ ]:
import pandas as pd

In [ ]:
df = pd.read_csv("/content/sample_data/SISAGRI.csv")

### Exportamos solo lo nesesario y que sea tratable para otros

In [ ]:
df_papa = df[df['PRODUCTO'].str.contains("papa ", case=False, na=False)]
df_papa = df_papa.drop(columns=['COD_UBIGEO', 'COD_PRODUCTO'])

In [ ]:
df_papa.to_csv('/content/sample_data/filtrado_papa.csv', index=False, sep=',', encoding='utf-8')
df = df_papa

## Exploración básica de los datos

In [ ]:
# 1. Ver las primeras 5 filas
print("------------------------Primeras 5 filas:------------------------")
print(df.head())

# 3. Resumen general del DataFrame
print("\n------------------------Información general del DataFrame:------------------------")
df.info()

# 7. Contar valores nulos por columna
print("\n------------------------Conteo de valores nulos y ceros:------------------------")
print((df.isnull() | (df == 0)).sum())
# 10. Contar frecuencia de valores en una columna (reemplazar 'COLUMNA' por el nombre de una columna)
print("\n------------------------Frecuencia de valores en una columna específica:------------------------")
print(df['PRODUCTO'].value_counts() if 'PRODUCTO' in df.columns else "La columna especificada no existe.")

------------------------Primeras 5 filas:------------------------
     AÑO  MES DEPARTAMENTO    PROVINCIA                DISTRITO  \
12  2015   12     AMAZONAS  CHACHAPOYAS             CHUQUIBAMBA   
50  2015    8     AMAZONAS         LUYA  SAN FRANCISCO DEL YESO   
57  2015    2     AMAZONAS         LUYA                   LAMUD   
58  2015    2     AMAZONAS         LUYA                   LAMUD   
72  2015    7    CAJAMARCA    CAJABAMBA               CAJABAMBA   

                             PRODUCTO  SIEMBRA  COSECHA  PRODUCCION  \
12       PAPA CHAUCHA (papas nativas)      0.0      1.0        15.0   
50        PAPA HUAIRO (papas nativas)      0.0      7.0       136.5   
57    PAPA CANCHAN (mejoradas)- color      3.0      0.0         0.0   
58            PAPA YUNGAY (mejoradas)      0.0      0.0         0.0   
72  PAPA (agrupa mejoradas y nativas)     40.0     33.0       165.3   

    PRECIO_CHACRA  
12           1.85  
50           0.87  
57           0.00  
58           0.00  
72  

#### Cantidad de filas que tienen 0 o null ya sea en Siembra o cosesha

In [ ]:
df[(df['SIEMBRA'].isnull() | (df['SIEMBRA'] == 0)) | (df['COSECHA'].isnull() | (df['COSECHA'] == 0))].shape[0]

55671

In [ ]:
df[(df['SIEMBRA'] < 0) | (df['COSECHA'] < 0)].shape[0]


0

In [ ]:
df_rend_sinPapasNoID = df[
    (df['SIEMBRA'].notnull() & (df['SIEMBRA'] != 0)) &
    (df['COSECHA'].notnull() & (df['COSECHA'] != 0)) &
    (~df['PRODUCTO'].isin([
        'PAPA (agrupa mejoradas y nativas)',
        'PAPA NATIVA (agrupa papas nativas)',
        'PAPA COLOR (agrupa papas mejoradas de color)',
        'PAPA MEJORADA (Agrupa papas mejoradas) '
    ]))
]
df_rend_sinPapasNoID

AÑO  MES DEPARTAMENTO    PROVINCIA          DISTRITO  \
1015     2015    8      HUANUCO     PACHITEA            MOLINO   
1491     2016    4     AMAZONAS         LUYA       SANTO TOMAS   
2494     2016    8      HUANUCO     PACHITEA            MOLINO   
2636     2016    3      HUANUCO      HUANUCO  QUISQUI (KICHKI)   
3797     2017    6      HUANUCO         AMBO     SAN FRANCISCO   
...       ...  ...          ...          ...               ...   
1044883  2017    8     AMAZONAS  CHACHAPOYAS            SOLOCO   
1045644  2018    4     AMAZONAS  CHACHAPOYAS         CHILIQUIN   
1046047  2022    4     AMAZONAS         LUYA    SANTA CATALINA   
1047069  2016    4     AMAZONAS  CHACHAPOYAS          ASUNCION   
1047866  2020    1     AMAZONAS         LUYA            CONILA   

                                PRODUCTO  SIEMBRA  COSECHA  PRODUCCION  \
1015       PAPA AMARILLA (papas nativas)     68.0     58.0       875.0   
1491         PAPA HUAIRO (papas nativas)     22.0     30.0       561.0   
2494       PAPA AMARILLA (papas nativas)     74.0    110.0      1683.0   
2636       PAPA AMARILLA (papas nativas)     15.0     65.0       877.0   
3797       PAPA AMARILLA (papas nativas)     13.0     20.0       241.0   
...                                  ...      ...      ...         ...   
1044883     PAPA CHAUCHA (papas nativas)      1.0      1.0        14.6   
1045644     PAPA CHAUCHA (papas nativas)      2.0      2.0        30.0   
1046047      PAPA HUAIRO (papas nativas)      1.0      4.0        80.0   
1047069      PAPA HUAIRO (papas nativas)      2.0      5.0        85.0   
1047866  PAPA CANCHAN (mejoradas)- color     38.0      1.0        18.0   

         PRECIO_CHACRA  
1015              1.00  
1491              0.86  
2494              1.60  
2636              0.86  
3797              1.02  
...                ...  
1044883           1.20  
1045644           1.00  
1046047           1.65  
1047069           0.70  
1047866           1.40  

[1218 rows x 10 columns]

In [ ]:
# Contar cuántos productos únicos hay
num_productos_unicos = df_rend_sinPapasNoID['PRODUCTO'].nunique()

# Obtener la lista de productos únicos con sus conteos
productos_unicos_conteo = df_rend_sinPapasNoID['PRODUCTO'].value_counts()

# Mostrar resultados
print(f"Cantidad de productos únicos: {num_productos_unicos}")
print("\nProductos y sus conteos:")
print(productos_unicos_conteo)


Cantidad de productos únicos: 12

Productos y sus conteos:
PRODUCTO
PAPA HUAIRO (papas nativas)                          399
PAPA YUNGAY (mejoradas)                              230
PAPA AMARILLA (papas nativas)                        176
PAPA CHAUCHA (papas nativas)                         174
PAPA CANCHAN (mejoradas)- color                       94
PAPA AMARILIS (mejoradas)                             85
PAPA UNICA (mejoradas)-color                          19
PAPA SUELA COLORADA (demás papas nativas)             17
PAPA PERRICHOLI (mejoradas)                           14
PAPA HUAMACHUQUINA O LIMEÑA (Demás papas nativas)      7
PAPA HUAMANTANGA (papas nativas)                       2
PAPA MARIVA (Demás papas mejoradas)                    1
Name: count, dtype: int64


In [ ]:
df_rend_sinPapasNoID_6 = df_rend_sinPapasNoID[
    df_rend_sinPapasNoID['PRODUCTO'].isin([
        'PAPA PERRICHOLI (mejoradas)',
        'PAPA CANCHAN (mejoradas)- color',
        'PAPA AMARILIS (mejoradas)',
        'PAPA UNICA (mejoradas)-color',
        'PAPA YUNGAY (mejoradas)',
        'PAPA HUAIRO (papas nativas)'
    ])
]

# Añadir columna 'RENDIMIENTO' calculada como COSECHA / SIEMBRA
df_rend_sinPapasNoID_6 = df_rend_sinPapasNoID_6.copy()  # Para evitar SettingWithCopyWarning
df_rend_sinPapasNoID_6['RENDIMIENTO'] = df_rend_sinPapasNoID_6['COSECHA'] / df_rend_sinPapasNoID_6['SIEMBRA']


In [ ]:
df_rend_sinPapasNoID_6

AÑO  MES  DEPARTAMENTO    PROVINCIA        DISTRITO  \
1491     2016    4      AMAZONAS         LUYA     SANTO TOMAS   
3847     2017    5  HUANCAVELICA     TAYACAJA      HUARIBAMBA   
8988     2021    4      AMAZONAS  CHACHAPOYAS     MOLINOPAMPA   
9471     2021    6      AMAZONAS  CHACHAPOYAS     CHACHAPOYAS   
9926     2021    9          LIMA       CAÑETE        SAN LUIS   
...       ...  ...           ...          ...             ...   
1042862  2022    3      AMAZONAS         LUYA           TRITA   
1044184  2017    5      AMAZONAS  CHACHAPOYAS       MAGDALENA   
1046047  2022    4      AMAZONAS         LUYA  SANTA CATALINA   
1047069  2016    4      AMAZONAS  CHACHAPOYAS        ASUNCION   
1047866  2020    1      AMAZONAS         LUYA          CONILA   

                                PRODUCTO  SIEMBRA  COSECHA  PRODUCCION  \
1491         PAPA HUAIRO (papas nativas)     22.0     30.0       561.0   
3847             PAPA YUNGAY (mejoradas)      6.0     25.0       300.0   
8988             PAPA YUNGAY (mejoradas)      5.0      2.0        35.0   
9471         PAPA HUAIRO (papas nativas)      3.0      2.0        35.0   
9926        PAPA UNICA (mejoradas)-color      3.0      4.0        96.0   
...                                  ...      ...      ...         ...   
1042862          PAPA YUNGAY (mejoradas)      1.0      1.0        22.0   
1044184      PAPA HUAIRO (papas nativas)      2.0      2.0        31.0   
1046047      PAPA HUAIRO (papas nativas)      1.0      4.0        80.0   
1047069      PAPA HUAIRO (papas nativas)      2.0      5.0        85.0   
1047866  PAPA CANCHAN (mejoradas)- color     38.0      1.0        18.0   

         PRECIO_CHACRA  RENDIMIENTO  
1491              0.86     1.363636  
3847              0.40     4.166667  
8988              0.60     0.400000  
9471              1.20     0.666667  
9926              0.45     1.333333  
...                ...          ...  
1042862           1.30     1.000000  
1044184           0.67     1.000000  
1046047           1.65     4.000000  
1047069           0.70     2.500000  
1047866           1.40     0.026316  

[841 rows x 11 columns]

In [ ]:
# Conteo de las clases en la columna PRODUCTO
conteo_productos = df_rend_sinPapasNoID_6['PRODUCTO'].value_counts()

# Conteo de los valores únicos de RENDIMIENTO, ordenados por los valores de RENDIMIENTO
conteo_rendimiento = df_rend_sinPapasNoID_6['RENDIMIENTO'].value_counts().sort_index()

# Imprimir los resultados
print("Conteo de las clases en la columna PRODUCTO:")
print(conteo_productos)

print("\nConteo de los valores únicos en la columna RENDIMIENTO (ordenados por los valores de RENDIMIENTO):")
print(conteo_rendimiento)


Conteo de las clases en la columna PRODUCTO:
PRODUCTO
PAPA HUAIRO (papas nativas)        399
PAPA YUNGAY (mejoradas)            230
PAPA CANCHAN (mejoradas)- color     94
PAPA AMARILIS (mejoradas)           85
PAPA UNICA (mejoradas)-color        19
PAPA PERRICHOLI (mejoradas)         14
Name: count, dtype: int64

Conteo de los valores únicos en la columna RENDIMIENTO (ordenados por los valores de RENDIMIENTO):
RENDIMIENTO
0.026316     1
0.041667     1
0.043478     1
0.058824     1
0.062500     1
            ..
13.333333    1
13.500000    1
14.000000    1
24.000000    1
33.000000    1
Name: count, Length: 167, dtype: int64


# CONLUSION
- en filtrado_papa.csv solo hay papas y sin columna de obigeos
	- contiene 61mil filas
	- tiene papas ambiguas
	-Produccion <- cosecha (depende)
	- tiene 55mil nulos o ceros en Siembra o Cosecha
	- no tiene valores negativos
- df_rend_sinPapasNoID:
	- 1218 filas
	- con papas semiambiguas ( sin normales,nativa,color,mejoradas) (como huairo,etc)
	- con cosecha y siembra >0
	-desproporcion masiva de clases
- df_rend_sinPapasNoID_6:
	- 841 filas
	-con las 6 que trabajmos
	-el rendimiento es: cosecha/siembra (nose xq la cosecha a veces sale mas xD)
	-reconsiderar entrar a la amarrilla y chaucha (300filas)
	- desproporcion de clases
	- rendimiento: [0.26-33]

# Sergio

### Nos quedamos solo con las papas

Valores de la tabla `papa_df`:


![image.png](attachment:image.png)


In [ ]:
papasdf = df[df["PRODUCTO"].str.startswith("PAPA")]
papasdf["RENDIMIENTO"] = papasdf["PRODUCCION"].div(papasdf["COSECHA"], fill_value=0)
papasdf = papasdf[papasdf["PRODUCTO"] != "PAPA (agrupa mejoradas y nativas)"]
# Incluyendo a los nulos
papasdf

<ipython-input-25-10e4c4da6ac1>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  papasdf["RENDIMIENTO"] = papasdf["PRODUCCION"].div(papasdf["COSECHA"], fill_value=0)


AÑO  MES  COD_UBIGEO DEPARTAMENTO          PROVINCIA  \
12       2015   12       10106     AMAZONAS        CHACHAPOYAS   
50       2015    8       10517     AMAZONAS               LUYA   
57       2015    2       10501     AMAZONAS               LUYA   
58       2015    2       10501     AMAZONAS               LUYA   
94       2015    2       10520     AMAZONAS               LUYA   
...       ...  ...         ...          ...                ...   
1047910  2020    5      130302  LA LIBERTAD            BOLIVAR   
1047934  2020    2      120120        JUNIN           HUANCAYO   
1047953  2023    9       50505     AYACUCHO             LA MAR   
1047965  2022    1      221005   SAN MARTIN            TOCACHE   
1047986  2018    9      131007  LA LIBERTAD  SANTIAGO DE CHUCO   

                       DISTRITO  COD_PRODUCTO  \
12                  CHUQUIBAMBA   14040090108   
50       SAN FRANCISCO DEL YESO   14040090113   
57                        LAMUD   14040090107   
58                        LAMUD   14040090127   
94               SANTA CATALINA   14040090113   
...                         ...           ...   
1047910              BAMBAMARCA   14040090103   
1047934             HUASICANCHA   14040090129   
1047953                 CHUNGUI   15010020000   
1047965                  UCHIZA   15010020000   
1047986     SANTA CRUZ DE CHUCA   14040090103   

                                             PRODUCTO  SIEMBRA  COSECHA  \
12                       PAPA CHAUCHA (papas nativas)      0.0      1.0   
50                        PAPA HUAIRO (papas nativas)      0.0      7.0   
57                    PAPA CANCHAN (mejoradas)- color      3.0      0.0   
58                            PAPA YUNGAY (mejoradas)      0.0      0.0   
94                        PAPA HUAIRO (papas nativas)     10.0      0.0   
...                                               ...      ...      ...   
1047910  PAPA COLOR (agrupa papas mejoradas de color)      0.0      9.0   
1047934            PAPA NATIVA (agrupa papas nativas)      0.0      0.0   
1047953                                        PAPAYA      0.0      0.0   
1047965                                        PAPAYA      0.0      0.0   
1047986  PAPA COLOR (agrupa papas mejoradas de color)      1.0      0.0   

         PRODUCCION  PRECIO_CHACRA  RENDIMIENTO  
12             15.0           1.85         15.0  
50            136.5           0.87         19.5  
57              0.0           0.00          NaN  
58              0.0           0.00          NaN  
94              0.0           0.00          NaN  
...             ...            ...          ...  
1047910       144.0           1.20         16.0  
1047934         0.0           0.00          NaN  
1047953         1.0           1.00          inf  
1047965        10.0           0.55          inf  
1047986         0.0           0.00          NaN  

[46950 rows x 13 columns]

In [ ]:
papasdf.to_csv('/content/sample_data/papas_filtradas.csv', index=False)


### Lista total de papas

No olvidar que nos quedamos con solo 6

In [ ]:
papasdf[papasdf["PRODUCTO"] != "PAPA (agrupa mejoradas y nativas)"] \
    .groupby("PRODUCTO") \
    .apply(lambda x: (x == 0).sum())


<ipython-input-26-05ca22a81366>:3: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  .apply(lambda x: (x == 0).sum())


AÑO  MES  COD_UBIGEO  \
PRODUCTO                                                                  
PAPA AMARILIS (mejoradas)                            0    0           0   
PAPA AMARILLA (papas nativas)                        0    0           0   
PAPA CANCHAN (mejoradas)- color                      0    0           0   
PAPA CHAUCHA (papas nativas)                         0    0           0   
PAPA COLOR (agrupa papas mejoradas de color)         0    0           0   
PAPA HUAIRO (papas nativas)                          0    0           0   
PAPA HUAMACHUQUINA O LIMEÑA (Demás papas nativas)    0    0           0   
PAPA HUAMANTANGA (papas nativas)                     0    0           0   
PAPA MARIVA (Demás papas mejoradas)                  0    0           0   
PAPA MEJORADA (Agrupa papas mejoradas)               0    0           0   
PAPA NATIVA (agrupa papas nativas)                   0    0           0   
PAPA PERRICHOLI (mejoradas)                          0    0           0   
PAPA SUELA COLORADA (demás papas nativas)            0    0           0   
PAPA UNICA (mejoradas)-color                         0    0           0   
PAPA YUNGAY (mejoradas)                              0    0           0   
PAPAS MEJORADAS Y NATIVAS                            0    0           0   
PAPAYA                                               0    0           0   
PAPAYUELA                                            0    0           0   

                                                   DEPARTAMENTO  PROVINCIA  \
PRODUCTO                                                                     
PAPA AMARILIS (mejoradas)                                     0          0   
PAPA AMARILLA (papas nativas)                                 0          0   
PAPA CANCHAN (mejoradas)- color                               0          0   
PAPA CHAUCHA (papas nativas)                                  0          0   
PAPA COLOR (agrupa papas mejoradas de color)                  0          0   
PAPA HUAIRO (papas nativas)                                   0          0   
PAPA HUAMACHUQUINA O LIMEÑA (Demás papas nativas)             0          0   
PAPA HUAMANTANGA (papas nativas)                              0          0   
PAPA MARIVA (Demás papas mejoradas)                           0          0   
PAPA MEJORADA (Agrupa papas mejoradas)                        0          0   
PAPA NATIVA (agrupa papas nativas)                            0          0   
PAPA PERRICHOLI (mejoradas)                                   0          0   
PAPA SUELA COLORADA (demás papas nativas)                     0          0   
PAPA UNICA (mejoradas)-color                                  0          0   
PAPA YUNGAY (mejoradas)                                       0          0   
PAPAS MEJORADAS Y NATIVAS                                     0          0   
PAPAYA                                                        0          0   
PAPAYUELA                                                     0          0   

                                                   DISTRITO  COD_PRODUCTO  \
PRODUCTO                                                                    
PAPA AMARILIS (mejoradas)                                 0             0   
PAPA AMARILLA (papas nativas)                             0             0   
PAPA CANCHAN (mejoradas)- color                           0             0   
PAPA CHAUCHA (papas nativas)                              0             0   
PAPA COLOR (agrupa papas mejoradas de color)              0             0   
PAPA HUAIRO (papas nativas)                               0             0   
PAPA HUAMACHUQUINA O LIMEÑA (Demás papas nativas)         0             0   
PAPA HUAMANTANGA (papas nativas)                          0             0   
PAPA MARIVA (Demás papas mejoradas)                       0             0   
PAPA MEJORADA (Agrupa papas mejoradas)                    0             0   
PAPA NATIVA (agrupa papas nativas)                        0      

### Lista de Papas por Departamento

In [ ]:
resultado = (papasdf.groupby(["DEPARTAMENTO", "PROVINCIA", "DISTRITO", "AÑO", "MES"])["PRODUCTO"]
        .nunique()
        .sort_values(ascending=False))

resultado[resultado == 3]